traffic_ops/traffic_ops_golang/deliveryservice/consistenthash/consistenthash.go

```diff
@@ -0,0 +1,134 @@
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	req := TRConsistentHashRequest{}
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("malformed JSON: "+err.Error()), nil)
+		return
+	}
+
+	responseFromTR, err := getPatternBasedConsistentHash(inf.Tx.Tx, req.ConsistentHashRegex, req.RequestPath, req.CdnID)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("getting pattern based consistent hash from Traffic Router: "+err.Error()))
+		return
+	}
+
+	consistentHashResult := TRConsistentHashResult{}
+	json.Unmarshal(responseFromTR, &consistentHashResult)
+
+	api.WriteResp(w, r, consistentHashResult)
+}
+
+const RouterRequestTimeout = time.Second * 10
+
+// queries database for active Traffic Router on the CDN specified by cdnId
+// passes regex and requestPath to the Traffic Router via API request,
+// and returns the response
+func getPatternBasedConsistentHash(tx *sql.Tx, regex string, requestPath string, cdnId int64) ([]byte, error) {
+	q := `
+SELECT concat(server.host_name, '.', server.domain_name) AS fqdn,
+   parameter.value AS apiport
+  FROM (((server
+    JOIN profile ON ((profile.id = server.profile)))
+    JOIN profile_parameter ON ((profile_parameter.profile = profile.id)))
+    JOIN parameter ON ((parameter.id = profile_parameter.parameter)))
+    JOIN status ON ((server.status = status.id))
+ WHERE ((server.type = (select id from type where name = 'CCR')) AND
+ (parameter.name = 'api.port'::text) AND
+ (status.name = 'ONLINE') AND
+ (server.cdn_id = $1))
+`
+	rows, err := tx.Query(q, cdnId)
+	defer rows.Close()
+	if err != nil {
+		return nil, errors.New("querying for eligible Traffic Router to test pattern based consistent hashing: " + err.Error())
+	}
+
+	trafficRouter := ""
+	apiPort := ""
+	for rows.Next() {
+		if err := rows.Scan(&trafficRouter, &apiPort); err != nil {
+			return nil, errors.New("scanning eligible Traffic Routers for pattern based consistent hashing: " + err.Error())
+		}
+	}
+
+	if trafficRouter == "" {
+		return nil, errors.New("no eligible Traffic Router found for pattern based consistent hashing with cdn Id: " + strconv.FormatInt(cdnId, 10))
+	}
+	if apiPort == "" {
+		return nil, errors.New("no parameter 'api.port' found for pattern based consistent hashing with cdn Id: " + strconv.FormatInt(cdnId, 10))
+	}
+
+	trafficRouterAPI := "http://" + trafficRouter + ":" + apiPort + "/crs/consistenthash/patternbased/regex?regex=" + url.QueryEscape(regex) + "&requestPath=" + url.QueryEscape(requestPath)
+
+	trClient := &http.Client{
+		Timeout: RouterRequestTimeout,
+	}
+	r, err := trClient.Get(trafficRouterAPI)
+
+	if err != nil {
+		return nil, errors.New("Error creating request to Traffic Router: " + err.Error())
+	}
+	defer r.Body.Close()
```

Comment:
We should also check to make sure the TR response is a 200, otherwise return an error like "Traffic Router returned a 404". When I was testing this locally it was making the request to a non-upgraded TR that didn't support the endpoint yet, and in Traffic Portal it wasn't showing an error.